src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -246,59 +236,79 @@ private static OtlpTrace.Status ToOtlpStatus(this Activity activity, ref TagEnum
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static OtlpTrace.Span.Types.Link ToOtlpLink(ActivityLink activityLink)
+        private static Span.Types.Link ToOtlpLink(in ActivityLink activityLink)
         {
             byte[] traceIdBytes = new byte[16];
             byte[] spanIdBytes = new byte[8];
 
             activityLink.Context.TraceId.CopyTo(traceIdBytes);
             activityLink.Context.SpanId.CopyTo(spanIdBytes);
 
-            var otlpLink = new OtlpTrace.Span.Types.Link
+            var otlpLink = new Span.Types.Link
             {
                 TraceId = UnsafeByteOperations.UnsafeWrap(traceIdBytes),
                 SpanId = UnsafeByteOperations.UnsafeWrap(spanIdBytes),
             };
 
-            TagEnumerationState otlpTags = default;
-            activityLink.EnumerateTags(ref otlpTags, SdkConfiguration.Instance.LinkAttributeCountLimit);
-            if (otlpTags.Created)
+            var enumerator = activityLink.EnumerateTagObjects();
+            if (enumerator.MoveNext())
             {
-                otlpLink.Attributes.AddRange(otlpTags.Tags);
-                otlpTags.Tags.Return();
+                int maxTags = SdkConfiguration.Instance.LinkAttributeCountLimit ?? int.MaxValue;
+                if (maxTags > 0)
+                {
+                    do
+                    {
+                        ref readonly var tag = ref enumerator.Current;
+                        if (OtlpKeyValueTransformer.Instance.TryTransformTag(tag, out var attribute, SdkConfiguration.Instance.AttributeValueLengthLimit))
+                        {
+                            otlpLink.Attributes.Add(attribute);
+                        }
+                    }
+                    while (enumerator.MoveNext() && otlpLink.Attributes.Count < maxTags);
```

Comment:
@alanwest If you look down below, for the top-level attributes, events, & links I updated the code to set the drop counts. Wasn't sure what to do for the sub-attributes, should those be reflected in those as well?

Reply:
events and links have their own "dropped count" field that should be populated. https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L210

Reply:
Thanks, updated!